File: works/1913.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import java.util.Collections;

public class Main
{


    public static void main(String[] args) throws IOException
    {
        Scanner sc = new Scanner(System.in);
        StringBuilder sb = new StringBuilder();

        int[] dr = {1, 0, -1, 0};
        int[] dc = {0, 1, 0, -1};
        int N = sc.nextInt();
        int num = sc.nextInt();
        int cnt = N*N;
        int r = 0, c = 0, dir = 0, ansR = 0, ansC = 0;
        int[][] v = new int[N][N];

        while(cnt > 0)
        {
            if(cnt == num) { ansR = r; ansC = c; }
            v[r][c] = cnt;
            int nr = r + dr[dir];
            int nc = c + dc[dir];

            if(nr < 0 || nr >= N || nc < 0 || nc >= N || v[nr][nc] != 0)
            {
                dir = (dir + 1) % 4;
                nr = r + dr[dir];
                nc = c + dc[dir];
            }
            r = nr; c = nc;
            cnt--;
        }



        for(int i = 0; i < N; i++)
        {
            for(int j = 0; j < N; j++)
            {
                sb.append(v[i][j]);
                sb.append(' ');
            } sb.append("\n");
        }
        sb.append(ansR+1); sb.append(' '); sb.append(ansC+1);
        System.out.print(sb);
    }
}

File: works/tmp/1913.kt
import java.util.*


fun main()
{
    val sc = Scanner(System.`in`)
    val sb = StringBuilder()
    val dr = arrayOf(1, 0, -1, 0)
    val dc = arrayOf(0, 1, 0, -1)
    var N = sc.nextInt()
    var num = sc.nextInt()
    var cnt = N*N
    var r = 0; var c = 0; var dir = 0; var ansR = 0; var ansC = 0
    var v = Array(N) {Array(N) {0} }

    while(cnt > 0)
    {
        if(cnt == num) {ansR= r; ansC = c}
        v[r][c] = cnt
        var nr = r + dr[dir]
        var nc = c + dc[dir]

        if (nr < 0 || nr >= N || nc < 0 || nc >= N || v[nr][nc] != 0) {
            dir = (dir + 1) % 4
            nr = r + dr[dir]
            nc = c + dc[dir]
        }
        r = nr; c = nc
        cnt--
    }

    for (i in 0 until N) {
        for (j in 0 until N) {
            sb.append(v[i][j])
            sb.append(' ')
        }
        sb.append("\n")
    }
    sb.append(ansR + 1); sb.append(' '); sb.append(ansC+1)
    print(sb)
}